k2/csrc/cuda/tensor.h

```diff
@@ -1,53 +1,67 @@
 // k2/csrc/cuda/tensor.h
 
-// Copyright (c)  2020  Xiaomi Corporation (authors: Daniel Povey)
+// Copyright (c)  2020  Xiaomi Corporation (authors: Daniel Povey
+//                                                   Haowen Qiu)
 
 // See ../../LICENSE for clarification regarding multiple authors
 
 #ifndef K2_CSRC_CUDA_TENSOR_H_
 #define K2_CSRC_CUDA_TENSOR_H_
 
+#include <vector>
+
 #include "k2/csrc/cuda/context.h"
 #include "k2/csrc/cuda/dtype.h"
 
-#define MAX_DIM 4  // Will increase this as needed
-
 namespace k2 {
 class Shape {
  public:
-  int32_t Ndim() { return ndim_; }
+  int32_t Ndim() const { return ndim_; }
+
+  std::vector<int32_t> Dims() const { return dims_; }
+
+  std::vector<int32_t> Strides() const { return strides_; }
 
-  int32_t Dim(int32_t i) {
+  int32_t Dim(int32_t i) const {
     CHECK_LT(static_cast<uint32_t>(i), static_cast<uint32_t>(ndim_));
     return dims_[i];
   }
-  int32_t Stride(int32_t i) {
+
+  int32_t Stride(int32_t i) const {
     CHECK_LT(static_cast<uint32_t>(i), static_cast<uint32_t>(ndim_));
     return strides_[i];
   }
-  int32_t Nelement();  // compute, return number of elements..
 
-  explicit Shape(const std::vector<int32_t> &dims, int32_t bytes_per_elem);
+  int32_t Nelement() const { return num_element_; }
+  bool IsContiguous() const { return is_contiguous_; }
+
+  Shape() : ndim_(0), num_element_(0), is_contiguous_(false) {}
+
+  explicit Shape(const std::vector<int32_t> &dims);
 
   explicit Shape(const std::vector<int32_t> &dims,
                  const std::vector<int32_t> strides);
 
-  Shape(const Shape &other);
-
-  bool IsContiguous(int32_t bytes_per_elem);
+  Shape(const Shape &other) = default;
 
  private:
+  static const int32_t kMaxDim = 4;  // Will increase this as needed
+
   int32_t ndim_;  // Must be >= 0
-  // elements of dims_ and strides_ >= ndim_ are currently not set;
-  // in future we may change this.
-  int32_t dims_[MAX_DIM];
-  int32_t strides_[MAX_DIM];  // Strides in bytes
+  int32_t num_element_;
```

Comment:
am asking Haowen to revert this change and do it as I drafted, with fixed maximum size (other toolkits do it like this + its more efficient)

Reply:
reverted